swagger/smart_transfer_protocol.yaml

```diff
@@ -0,0 +1,230 @@
+openapi: 3.0.3
+
+info:
+  title: ODF Smart Transfer Protocol
+  description:
+    Open Data Fabric (ODF) extended protocol for dataset exchange between remote parties.
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: 0.0.1
+
+
+components:
+
+  parameters:
+
+    DatasetId:
+      in: path
+      name: datasetId
+      description: Dataset identifier
+      example: did:odf:z4k88e8rxU6m5wCnK9idM5sGAxAGfvUgNgQbckwJ4ro78tXMLSu
+      required: true
+      schema:
+        $ref: '#/components/schemas/DatasetId'
+
+    StartBlockHash:
+      in: query
+      name: startBlockHash
+      description: Hash code of the starting block
+      example: zW1qJPmDvBxGS9GeC7PFseSCy7koHjvurUmisf1VWscY3AX
+      required: true
+      schema:
+        $ref: '#/components/schemas/MultiHash'
+
+    TailBlockHash:
+      in: query
+      name: tailBlockHash
+      description: Hash code of the tail block
+      example: zW1qJPmDvBxGS9GeC7PFseSCy7koHjvurUmisf1VWscY3AX
+      required: true
+      schema:
+        $ref: '#/components/schemas/MultiHash'
+
+    BlocksCount:
+      in: query
+      name: blocksCount
+      description: Max number of chained blocks to process. If not specified, unlimited.
+      example: 5
+      required: false
+      schema:
+        type: integer
+        format: int32
+
+  schemas:
+
+    DatasetId:
+      type: string
+      format: did
+
+    MultiHash:
+      type: string
+      format: multihash
+
+    ListOfBlocks:
+      type: array
+      items:
+        $ref: '../schemas/MetadataBlock.json'
+
+    ListOfCheckpointInfos:
+      type: array
+      items:
+        $ref: '#/components/schemas/CheckpointInfo'
+
+    CheckpointInfo:
+      type: object
+      required:
+        - checkpoint
+        - storage
+      properties:
+        checkpoint:
+          $ref: '../schemas/fragments/Checkpoint'
+        storage:
+          $ref: '#/components/schemas/StorageReference'
+
+    ListOfDataSliceInfos:
+      type: array
+      items:
+        $ref: '#/components/schemas/DataSliceInfo'
+
+    DataSliceInfo:
+      type: object
+      required:
+        - slice
+        - storage
+      properties:
+        slice:
+          $ref: '../schemas/fragments/DataSlice'
+        storage:
+          $ref: '#/components/schemas/StorageReference'
+
+    StorageReference:
+      type: object
+      required:
+        - provider
+        - url
+      properties:
+        provider:
+          type: string
+          enum: [S3, GCS, ADLS]
+        url:
+          type: string
+          format: url
+
+  responses:
+    Error:
+      description: Request error
+      content:
+        application/json:
+          schema:
+            type: object
+
+  securitySchemes:
+    odf_access_token:
+      type: http
+      scheme: bearer
+      bearerFormat: org.odf.access_token
+
+paths:
+
+  /datasets/{datasetId}/head:
```

Comment:
I have an idea:

Technically speaking our `Simple Transfer Protocol` already specifies a REST API spec that looks something like this:
- `GET /refs/{name} -> text/plain`
- `GET /blocks/{hash} -> application/octet-stream`
- `GET /data/{physicalHash} -> application/octet-stream`
- `GET /checkpoints/{physicalHash} -> application/octet-stream`

So how about we make `Smart Transfer Protocol` a **superset** of `Simple T.P.`, basically by adding two endpoints:
- `GET /pull/{name} Connection: upgrade; Upgrade: websocket; Sec-WebSocket-Protocol: open-data-fabric`
- `GET /push/{name} Connection: upgrade; Upgrade: websocket; Sec-WebSocket-Protocol: open-data-fabric`

This will allow clients that don't want to use Websocket still talk to Smart Protocol servers as if they were pure storage themselves.

What I also like about this that the **URL path scheme** is super liberal - it's up to the server to decide how to identify which dataset the request belongs to - it can use any prefix:
- `/{account}/{dataset_name}`
- `/datasets/{dataset_id}`
- whatever

If we move this "identity" concern outside of the spec if simplifies protocol significantly - we don't even need `DatasetID` as a parameter in any of the messages!

Reply:
That's fine, not changing the game for the API much. Yes, we could remove `datasetId` from `DatasetPullRequest` and `DatasetPushRequest` messages, and start the co-routine with the dataset ID passed via arguments. The rest of the protocol is unaffected.

Reply:
Added Swagger part